Retrospective on the May-July 2024 roadmap (#59)
linebender/linebender.github.io: content/blog/2024-08-23-roadmap-may-2024-retro.md

### Progress beyond roadmap {#vello-extras}

We have also landed several important features which were not part of our planned roadmap.
These include:

- Blending which uses more than four deep layers ([vello#661](https://github.com/linebender/vello/pull/661)).
- Preliminary support for embedding pre-existing `wgpu` textures ([vello#636](https://github.com/linebender/vello/pull/636)).
  This was driven by user needs.
  Further work in this direction is tracked in [vello#664](https://github.com/linebender/vello/issues/664).
- Support for Emoji (COLRv1 in [vello#615](https://github.com/linebender/vello/pull/615), bitmap in [vello#641](https://github.com/linebender/vello/pull/641)).

The [stroke expansion paper](https://linebender.org/gpu-stroke-expansion-paper/) documents a key part of Vello's pipeline.
This was presented at ACM High Performance Graphics (HPG) 2024, placing 3rd Place in the Best Paper Award.

## Masonry [(original)](@/blog/2024-06-17-roadmap-may-2024.md#masonry)

### Complete text input [(original)](@/blog/2024-06-17-roadmap-may-2024.md#complete-text-input)

- **Priority**: Highest
- **Status**: Not that much progress.

>   There were some community contributions such as

Reviewer: incomplete sentence?

### Better tracing for debugging [(original)](@/blog/2024-06-17-roadmap-may-2024.md#better-tracing-for-debugging)

- **Priority**: High
- **Status**: [xilem#384](https://github.com/linebender/xilem/pull/384) implements most of this, the rest is deferred.
  There are known issues with the size of the log files created by default in very large apps.

### Fix glaring documentation problems [(original)](@/blog/2024-06-17-roadmap-may-2024.md#fix-glaring-documentation-problems)

- **Priority**: High

> - **Status**: The issues mentioned in this item were addressed in [xilem#329](https://github.com/linebender/xilem/pull/329), and.

Reviewer: incomplete sentence?

### Finish repository port [(original)](@/blog/2024-06-17-roadmap-may-2024.md#fix-glaring-documentation-problems)

- **Priority**: High
- **Status**: This was completed.
  We moved all Masonry issues into the Xilem repository, and updated all references in code to use full links.
  This move was coordinated in [#linebender > Moving Masonry into Xilem](https://xi.zulipchat.com/#narrow/stream/419691-linebender/topic/Moving.20Masonry.20into.20Xilem).

### Write “widgets in arena” RFC [(original)](@/blog/2024-06-17-roadmap-may-2024.md#write-widgets-in-arena-rfc)

- **Priority**: High
- **Status**: The RFC was [rfcs#6](https://github.com/linebender/rfcs/pull/6), which was mostly implemented in [xilem#396](https://github.com/linebender/xilem/pull/396).

### Write “pass order” RFC [(original)](@/blog/2024-06-17-roadmap-may-2024.md#write-pass-order-rfc)

- **Prioriy**: High
- **Status**: [rfcs#7](https://github.com/linebender/rfcs/pull/7) is in draft, and several important parts of the RFC have been completed.

>   We have also learned several important lessons about the RFC process, including.

Reviewer: incomplete sentence?

### Rewrite documentation [(original)](@/blog/2024-06-17-roadmap-may-2024.md#rewrite-documentation)

- **Priority**: Medium
- **Status**: This work has been deferred until the "pass order" refactor is complete, as we expect Masonry's core to be quite stable at that point.

### Improve focus handling [(original)](@/blog/2024-06-17-roadmap-may-2024.md#improve-focus-handling) ?

- **Priority**: Medium
- **Status**: Not done.

### Improve pointer status handling [(original)](@/blog/2024-06-17-roadmap-may-2024.md#improve-pointer-status-handling) ?

- **Priority**: Medium
- **Status**: Basic translation of touch events was added in ([xilem#313](https://github.com/linebender/xilem/pull/313)).
  This unblocked using Masonry on devices where pointer input is not expected (primarily on Android).
  We have split the remainder of this task into two tasks, with a priority placed on completing the parts which unblock other work (such as scrolling on Android).

### Progress beyond roadmap {#masonry-extras}

Masonry is in a state of flux, however is starting to see a few exploratory users.
To unblock these users, some additional important features were added outside of the roadmap, such as:

> - Support for embedding Masonry inside a pre-existing event loop ([xilem#417](https://github.com/linebender/xilem/pull/417)).

Reviewer 2: The link is a 404. I *think* you're referring to [xilem#325](https://github.com/linebender/xilem/pull/325)?

Author: No, that link is correct

GitHub has deleted the pull request, but the change I'm referring to happened in that PR.

Author: I've linked to https://github.com/linebender/xilem/commit/cf3530097baf0fc5c458593e215a19973572b991 instead.

- **Priority**: High
- **Status**: Some work was done towards this in [parley#26](https://github.com/linebender/parley/pull/26), which unfortunately could not be merged.
  [parley#97](https://github.com/linebender/parley/pull/97) added some documentation to Fontique.
  Further documentation work from Nico is currently blocked on the tree-like styles PR [parley#76](https://github.com/linebender/parley/pull/76) being merged (to avoid conflicts).

### Software renderer [(original)](@/blog/2024-06-17-roadmap-may-2024.md#software-renderer)

- **Priority**: Low
- **Status**: This was not completed in the form envisioned.
  However, Nico contributed two examples, showing how to use Parley with Swash ([parley#54](https://github.com/linebender/parley/pull/54)) and tiny-skia ([parley#55](https://github.com/linebender/parley/pull/55)).
  We have decided not to carry this forward into the next roadmap.

### Progress beyond roadmap {#parley-extras}

[parley#56](https://github.com/linebender/parley/pull/56) added support for Emoji, which is used with the [Vello PRs](#vello-extras).

> <!-- TODO: Accurate sentence on Harfruzz -->

Reviewer 3: Here's my proposed sentence:

Recently the [harfruzz] project, a fork of rustybuzz to be integrated fully with fontations, has emerged as an extremely promising direction for text shaping for the Rust ecosystem. We plan to use it as the basis for text layout.

- **Priority**: Low
- **Status**: There has been no progress on this item.
  This will be a "highest" priority item in the next roadmap.

### Progress beyond roadmap {#xilem-extras}

Xilem has also seen some very useful changes not tracked above, including many community contributions.
Some highlights of these contributions are:

- Philipp Mildenberger contributed heavily to Xilem's core design, such as by restoring the ability to return `ViewSequence`s by returning to a previous workaround for Rust's overly-cautious trait solver ([xilem#472](https://github.com/linebender/xilem/pull/394)).
  He also extended the Xilem API by restoring `OneOf` ([xilem#394](https://github.com/linebender/xilem/pull/394)), and implementing various `Adapt` nodes ([xilem#401](https://github.com/linebender/xilem/pull/401)).
  Additionally, he picked up the work to expand `Flex` to use the [new Xilem Core](#rework-core) in [xilem#428](https://github.com/linebender/xilem/pull/428).
- Jared O'Connell contributed two new examples, a calculator ([xilem#467](https://github.com/linebender/xilem/pull/467)), and a stopwatch ([xilem#492](https://github.com/linebender/xilem/pull/492)).

> There has also been lots of really good progress on Xilem Web, but we do not roadmap Xilem Web and so is not discussed here.

Reviewer 2: ```suggestion
There has also been lots of really good progress on Xilem Web, but we do not roadmap Xilem Web and so we do not cover that progress here.
```
